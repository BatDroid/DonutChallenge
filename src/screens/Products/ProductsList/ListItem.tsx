import React from 'react';
import {
  View,
  Text,
  StyleSheet,
  Dimensions,
  TouchableHighlight,
} from 'react-native';
import {ProductType} from '../../../redux/products/actions/types';

const {height} = Dimensions.get('window');
export const ITEM_HEIGHT = height * 0.068;

interface Props {
  item: ProductType;
  onItemPressed: (product: ProductType) => void;
}

export default ({item, onItemPressed}: Props) => {
  const {display_name, status} = item;
  const onPress = () => onItemPressed(item);
  return (
    <TouchableHighlight onPress={onPress} underlayColor="#cccc">
      <View style={styles.itemRoot}>
        <View style={styles.infoContainer}>
          <Text style={styles.displayName}>{display_name}</Text>
          <Text style={styles.status}>{status}</Text>
        </View>
        {/* todo: add chevron Icon from react-native-vector-icons */}
        <Text style={styles.chevron}>></Text>
      </View>
    </TouchableHighlight>
  );
};

const styles = StyleSheet.create({
  itemRoot: {
    height: ITEM_HEIGHT,
    borderBottomWidth: 0.5,
    borderColor: '#cccc',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  infoContainer: {
    justifyContent: 'space-evenly',
    marginLeft: '2%',
  },
  displayName: {
    fontSize: 20,
  },
  chevron: {
    fontSize: 20,
    marginRight: '2%',
  },
  status: {
    color: '#666',
  },
});
