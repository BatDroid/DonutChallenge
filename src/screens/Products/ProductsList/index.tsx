import React from 'react';
import {FlatListProps, FlatList} from 'react-native';
import ListItem, {ITEM_HEIGHT} from './ListItem';
import {ProductType} from 'src/redux/products/actions/types';

interface Props extends Omit<FlatListProps<ProductType>, 'renderItem'> {
  onItemPressed: (prodcut: ProductType) => void;
}

export default (props: Props) => {
  const {onItemPressed, ...listProps} = props;
  return (
    <FlatList
      keyExtractor={i => i.id.toString()}
      initialNumToRender={18}
      renderItem={({item}) => {
        return <ListItem item={item} onItemPressed={onItemPressed} />;
      }}
      getItemLayout={(_, index) => ({
        length: ITEM_HEIGHT,
        offset: ITEM_HEIGHT * index,
        index,
      })}
      {...listProps}
    />
  );
};
