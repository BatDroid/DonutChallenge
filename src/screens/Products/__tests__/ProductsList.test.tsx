import 'react-native';
import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {render, fireEvent} from 'react-native-testing-library';
import ProductsList from '../ProductsList';

const mockProducts = [
  {
    id: 'EOS-BTC',
    base_currency: 'EOS',
    quote_currency: 'BTC',
    base_min_size: '0.10000000',
    base_max_size: '5000.00000000',
    quote_increment: '0.00000100',
    base_increment: '0.10000000',
    display_name: 'EOS/BTC',
    min_market_funds: '0.001',
    max_market_funds: '30',
    margin_enabled: false,
    post_only: false,
    limit_only: true,
    cancel_only: false,
    status: 'online',
    status_message: '',
  },
];

describe('Products List', () => {
  const onItemPressed = jest.fn();

  it('renders correctly', () => {
    const Component = shallow(
      <ProductsList data={mockProducts} onItemPressed={onItemPressed} />,
    );
    expect(toJson(Component)).toMatchSnapshot();
  });

  it('returns the pressed Item data when it is pressed', () => {
    const {getByText} = render(
      <ProductsList data={mockProducts} onItemPressed={onItemPressed} />,
    );
    const selectedItem = mockProducts[0];
    fireEvent.press(getByText(selectedItem.display_name));
    expect(onItemPressed).toHaveBeenCalledWith(selectedItem);
  });
});
