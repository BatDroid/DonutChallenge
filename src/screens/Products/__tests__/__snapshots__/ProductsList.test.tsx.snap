// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products List renders correctly 1`] = `
<FlatList
  data={
    Array [
      Object {
        "base_currency": "EOS",
        "base_increment": "0.10000000",
        "base_max_size": "5000.00000000",
        "base_min_size": "0.10000000",
        "cancel_only": false,
        "display_name": "EOS/BTC",
        "id": "EOS-BTC",
        "limit_only": true,
        "margin_enabled": false,
        "max_market_funds": "30",
        "min_market_funds": "0.001",
        "post_only": false,
        "quote_currency": "BTC",
        "quote_increment": "0.00000100",
        "status": "online",
        "status_message": "",
      },
    ]
  }
  disableVirtualization={false}
  getItemLayout={[Function]}
  horizontal={false}
  initialNumToRender={18}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
